marlowe-dashboard-client/src/Marlowe/Execution/State.purs

```diff
@@ -8,54 +8,53 @@ module Marlowe.Execution.State
   , getActionParticipant
   , extractNamedActions
   , expandBalances
+  , getAllPayments
   ) where
 
 import Prelude
 import Data.Array as Array
 import Data.BigInteger (fromInt)
 import Data.Lens (view, (^.))
-import Data.List (List)
+import Data.List (List, concat, fromFoldable, snoc)
 import Data.Map as Map
 import Data.Maybe (Maybe(..), fromMaybe, fromMaybe')
 import Data.Tuple.Nested ((/\))
+import Marlowe.Execution.Lenses (_currentPayments, _payments)
 import Marlowe.Execution.Types (ExecutionState, NamedAction(..), PendingTimeouts)
-import Marlowe.Semantics (Accounts, Action(..), Case(..), ChoiceId(..), Contract(..), Input, Party, ReduceResult(..), Slot(..), SlotInterval(..), State, Timeouts(..), Token, TransactionInput(..), TransactionOutput(..), _accounts, _boundValues, _minSlot, computeTransaction, emptyState, evalValue, makeEnvironment, reduceContractUntilQuiescent, timeouts)
+import Marlowe.Semantics (Accounts, Action(..), Case(..), ChoiceId(..), Contract(..), Input, Party, Payment, ReduceResult(..), Slot(..), SlotInterval(..), State, Timeouts(..), Token, TransactionInput(..), TransactionOutput(..), _accounts, _boundValues, _minSlot, computeTransaction, emptyState, evalValue, makeEnvironment, reduceContractUntilQuiescent, timeouts)
 
 initExecution :: Slot -> Contract -> ExecutionState
 initExecution currentSlot contract =
-  let
-    previous = mempty
-
-    state = emptyState currentSlot
-  in
-    { previous
-    , current:
-        { state
-        , contract
-        }
-    , mPendingTimeouts: Nothing
-    , mNextTimeout: nextTimeout contract
-    }
+  { previousExecutionStates: mempty
+  , currentExecutionState:
+      { state: emptyState currentSlot
+      , contract
+      , payments: mempty
+      }
+  , mPendingTimeouts: Nothing
+  , mNextTimeout: nextTimeout contract
+  }
 
 -- FIXME: Change the order of the arguments to::  TransactionInput -> ExecutionState  -> ExecutionState
 nextState :: ExecutionState -> TransactionInput -> ExecutionState
-nextState { previous, current } txInput =
+nextState { previousExecutionStates, currentExecutionState } txInput =
```

Comment:
not quite sure about this renaming here... Why would `ExecutionState` include a previous and current `ExecutionState`? This implies recursion (which is not the case). I think `previous` and `current` work just fine as they are, because you have:

```
ExecutionState = {current.state + current.contract + previous[0].state + etc}
```

Reply:
Okay, so my thinking is we've got `previous :: Array PreviousState` and `current :: { ... }`, so I naturally think of `current` as imlicitly having the type `CurrentState` (even though it doesn't have an explicit name). But both of these things have a `state :: State` property, so that's weird. `PreviousState.State` and `CurrentState.State` feel wrong. Also the lens `_currentState` feels like a bit of a cheat - implicitly it feels like it should be `_currentStateState`, but that's clearly horrible.

So I wanted other names a bit more distinctive than `PreviousState` and `CurrentState`, and I couldn't think of anything better than `PreviousExecutionState` and `CurrentExecutionState` - which I agree aren't great. :confused: 

I just now thought of `PreviousContractState` and `CurrentContractState`. How do you feel about those?

Reply:
mhmhmh, the thing is that they have more information than just the contract state.

Reply:
For example we currently have:

`previous[ix].contract`, `previous[ix].txInput` and `previous[ix].state` (which is the contract state/core state/execution state/semantic state).